### ISSUE TITLE
Knyst on android

### ISSUE THREAD

**Issue by Tuurlijk** [ErikNatanael/knyst#26]
Hi,

I would love to use knyst in an android app. I have a working example for using cpal here:

https://github.com/Tuurlijk/android-cpal-test/tree/main

And a similar test for knysq here

https://github.com/Tuurlijk/android-cpal-test/tree/dev/knyst

The cpal one plays a tone on device (native), not on the simulator. The knyst example errors out on both device and simulator.

Am I doing something wrong?

**Comment by ErikNatanael** [comment]
Hi!
Glad you're trying out Knyst on Android. I'm unable to run your example, could you post the errors you're getting and/or more instructions for how to run the knyst example?

**Comment by Tuurlijk** [comment]
I updated the readme. It comes down to:

## How to run

Build the libs with the build script: `./build.sh`

Connect your phone in usb debugging mode.

Then build the android app in Android Studio.

Or you can use gradlew and adb:

```shell
./gradlew build
./gradlew installDebug
```
The `installDebug` command will install the apk on the active devices.

Then run the app:
```shell
adb shell am start -n com.example.androidcpaltest/.MainActivity
```

If you have multiple devices attached, you can list them with: `adb devices`. Then target a device to run the activity on:

```shell
 adb devices
List of devices attached
c0ad78fa	device
emulator-5554	device
```

```shell
adb -s c0ad78fa shell am start -n com.example.androidcpaltest/.MainActivity
```

**Comment by Tuurlijk** [comment]
Error from emulator:

```
--------- beginning of main
--------- beginning of system
--------- beginning of crash
11:23:02.930  I  Late-enabling -Xcheck:jni
11:23:02.964  E  Unknown bits set in runtime_flags: 0x8000
11:23:02.965  W  Unexpected CPU variant for X86 using defaults: x86
11:23:04.520  D  Emulator has host GPU support, qemu.gles is set to 1.
11:23:04.520  W  Unable to set property "qemu.gles" to "1": connection failed; errno=13 (Permission denied)
11:23:04.511  W  type=1400 audit(0.0:251): avc: denied { write } for name="property_service" dev="tmpfs" ino=7725 scontext=u:r:untrusted_app:s0:c137,c256,c512,c768 tcontext=u:object_r:property_socket:s0 tclass=sock_file permissive=0
11:23:04.545  D  loaded /vendor/lib/egl/libEGL_emulation.so
11:23:04.547  D  loaded /vendor/lib/egl/libGLESv1_CM_emulation.so
11:23:04.550  D  loaded /vendor/lib/egl/libGLESv2_emulation.so
11:23:04.880  A  Fatal signal 11 (SIGSEGV), code 2 (SEGV_ACCERR), fault addr 0xff201ffc in tid 24917 (androidcpaltest), pid 24917 (androidcpaltest)
11:23:04.965  A  pid: 24917, tid: 24917, name: androidcpaltest  >>> com.example.androidcpaltest <<<
11:23:05.084  A        #00 pc 00150db7  /data/app/com.example.androidcpaltest-IS5Q1buDyQvScH3gtGfQjQ==/base.apk (offset 0xc78000)
11:23:05.084  A        #01 pc 0009a166  /data/app/com.example.androidcpaltest-IS5Q1buDyQvScH3gtGfQjQ==/base.apk (offset 0xc78000)

```

**Comment by Tuurlijk** [comment]
You may also need to set these env vars so ndk knows where to read from:

```shell
export ANDROID_HOME=${HOME}/Applications/Android/Sdk
export ANDROID_SDK_ROOT=${HOME}/Applications/Android/Sdk
export ANDROID_NDK_ROOT=${ANDROID_SDK_ROOT}/ndk/28.0.12433566
export ANDROID_NDK_HOME=${HOME}/Applications/Android/Sdk/ndk/28.0.12433566
export ANDROID_API=29
```

**Comment by ErikNatanael** [comment]
My initial thought is that the threads on Android could be more sandboxed than on other platforms, I don't know how this code interacts with Java and the JVM. The CPAL and JACK backends start a new thread behind the scenes which updates the running graph, and this thread "owns" the memory that is being accessed in the audio callback. In your CPAL example above, the audio callback owns all the memory it needs. 

It would be very helpful to narrow down exactly where the segfault occurs. Do you know if it's possible to run the app through a debugger and get a backtrace to the Rust code?

**Comment by Tuurlijk** [comment]
I'll try to get debugging working.

**Comment by Tuurlijk** [comment]
I can step through the execution of the cpal example (also see the calls being done inside the rust library), but as soon as I try to do the same with the knyst example, the app crashes. It leaves no `tombstone` file. Or I don't know where it dumps it. I looked inside the virtual phone and the `/data/tombstones`.

So I fallback to deduction mode. I commented out all the knyst code and it still crashes. Comment out all the use statements and it runs. Add back any use from the knyst prelude and it crashes again. 

I link against a local knyst repo, so my next course of action would be to 'undress' knyst bit by bit untill I find the breaking thing.

But maybe there is something I am overlooking here. Or something with the many macro's in use?
